<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>process_BC_data &mdash; p_ssGSEA 1 documentation</title>
    
    <link rel="stylesheet" href="../_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="top" title="p_ssGSEA 1 documentation" href="../index.html" />
    <link rel="up" title="Module code" href="index.html" />
   
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head>
  <body role="document">  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <h1>Source code for process_BC_data</h1><div class="highlight"><pre>
<span></span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">This file contains methods for importing and pre-processing the data from the Breast Cancer data set used in the</span>
<span class="sd">paper &quot;Evaluating gene set enrichment analysis via a hybrid data model&quot; by J. Hua et al. 2014.</span>

<span class="sd">ASSUMES ALL FILES ARE ENCODED AS UTF-7! (Which they are originally!)</span>
<span class="sd">&#39;&#39;&#39;</span>

<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">glob</span>

<span class="kn">from</span> <span class="nn">data_models</span> <span class="kn">import</span> <span class="o">*</span>

<span class="n">BC_DATA_DIR</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">getcwd</span><span class="p">()</span> <span class="o">+</span> <span class="s2">&quot;/Data/HybridSets/BC&quot;</span>
<span class="n">BC_EXPRESSION_DIR</span> <span class="o">=</span> <span class="n">BC_DATA_DIR</span> <span class="o">+</span> <span class="s2">&quot;/ExpressionProfiles&quot;</span>
<span class="n">BC_CLINICAL_DATA_FILE</span> <span class="o">=</span> <span class="n">BC_DATA_DIR</span> <span class="o">+</span> <span class="s2">&quot;/ClinicalData/ClinicalData.txt&quot;</span>


<div class="viewcode-block" id="BC_FileFormatError"><a class="viewcode-back" href="../process_BC_data.html#process_BC_data.BC_FileFormatError">[docs]</a><span class="k">class</span> <span class="nc">BC_FileFormatError</span><span class="p">(</span><span class="ne">Exception</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    This class represents an error in formatting from a file from the BC data set.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="nb">file</span><span class="p">,</span> <span class="n">error</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">file</span> <span class="o">=</span> <span class="nb">file</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">error</span> <span class="o">=</span> <span class="n">error</span>

    <span class="k">def</span> <span class="nf">__str__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="s2">&quot;File &quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">file</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">error</span><span class="p">)</span></div>

<div class="viewcode-block" id="readExpressionProfile"><a class="viewcode-back" href="../process_BC_data.html#process_BC_data.readExpressionProfile">[docs]</a><span class="k">def</span> <span class="nf">readExpressionProfile</span><span class="p">(</span><span class="n">file_name</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Reads a single file from the BC set containing gene expression information.</span>

<span class="sd">    :param file_name: the path and file containing expression information to read</span>
<span class="sd">    :type file_name: str</span>

<span class="sd">    :return: a list of expression_profiles</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="nb">file</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">file_name</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">,</span> <span class="n">encoding</span><span class="o">=</span><span class="s2">&quot;utf-7&quot;</span><span class="p">)</span>

    <span class="c1">#get rows of the file</span>
    <span class="n">rows</span> <span class="o">=</span> <span class="nb">file</span><span class="o">.</span><span class="n">read</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">if</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">rows</span><span class="p">)</span> <span class="o">&lt;</span> <span class="mi">2</span><span class="p">):</span>
        <span class="k">raise</span> <span class="n">BC_FileFormatError</span><span class="p">(</span><span class="nb">file</span><span class="p">,</span> <span class="s2">&quot;Labels in top two rows missing&quot;</span><span class="p">)</span>

    <span class="c1">#extract sample id&#39;s from rows</span>
    <span class="n">IDs</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">first_row</span> <span class="o">=</span> <span class="n">rows</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">first_row</span><span class="p">[:]</span> <span class="o">=</span> <span class="p">(</span><span class="n">val</span> <span class="k">for</span> <span class="n">val</span> <span class="ow">in</span> <span class="n">first_row</span> <span class="k">if</span> <span class="n">val</span> <span class="o">!=</span> <span class="s1">&#39;&#39;</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">string</span> <span class="ow">in</span> <span class="n">first_row</span><span class="p">:</span>
        <span class="k">if</span> <span class="s2">&quot;Sample&quot;</span> <span class="ow">in</span> <span class="n">string</span><span class="p">:</span>
            <span class="n">IDs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">string</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]))</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">BC_FileFormatError</span><span class="p">(</span><span class="nb">file</span><span class="p">,</span> <span class="s2">&quot;Sample IDs expected or incorrectly formatted&quot;</span><span class="p">)</span>

    <span class="c1">#extract expression data into here</span>
    <span class="n">expression_profiles</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="c1">#start from 2 which is the first row that contains gene expression data</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">rows</span><span class="p">)):</span>
        <span class="n">row</span> <span class="o">=</span> <span class="n">rows</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>

        <span class="c1">#some of the data has a random newline</span>
        <span class="k">if</span> <span class="nb">str</span><span class="p">(</span><span class="n">row</span><span class="p">)</span> <span class="o">==</span> <span class="s1">&#39;&#39;</span><span class="p">:</span>
            <span class="k">break</span>

        <span class="c1">#each row contains multiple samples for each gene45</span>
        <span class="n">fields</span> <span class="o">=</span> <span class="n">row</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="n">substance</span> <span class="o">=</span> <span class="n">fields</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">gene</span> <span class="o">=</span> <span class="n">fields</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>

        <span class="c1">#start from two, which is the index of the first column with actual information</span>
        <span class="c1">#5 fields, so offset by 5 with each id</span>
        <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">IDs</span><span class="p">)):</span>
            <span class="n">sample_num</span>  <span class="o">=</span> <span class="n">IDs</span><span class="p">[</span><span class="n">f</span><span class="p">]</span>
            <span class="n">log_ratio</span>   <span class="o">=</span> <span class="n">fields</span><span class="p">[</span><span class="mi">2</span> <span class="o">+</span> <span class="mi">5</span> <span class="o">*</span> <span class="n">f</span> <span class="o">+</span> <span class="mi">0</span><span class="p">]</span>
            <span class="n">log_error</span>   <span class="o">=</span> <span class="n">fields</span><span class="p">[</span><span class="mi">2</span> <span class="o">+</span> <span class="mi">5</span> <span class="o">*</span> <span class="n">f</span> <span class="o">+</span> <span class="mi">1</span><span class="p">]</span>
            <span class="n">p_value</span>     <span class="o">=</span> <span class="n">fields</span><span class="p">[</span><span class="mi">2</span> <span class="o">+</span> <span class="mi">5</span> <span class="o">*</span> <span class="n">f</span> <span class="o">+</span> <span class="mi">2</span><span class="p">]</span>
            <span class="n">intensity</span>   <span class="o">=</span> <span class="n">fields</span><span class="p">[</span><span class="mi">2</span> <span class="o">+</span> <span class="mi">5</span> <span class="o">*</span> <span class="n">f</span> <span class="o">+</span> <span class="mi">3</span><span class="p">]</span>
            <span class="n">flag</span>        <span class="o">=</span> <span class="n">fields</span><span class="p">[</span><span class="mi">2</span> <span class="o">+</span> <span class="mi">5</span> <span class="o">*</span> <span class="n">f</span> <span class="o">+</span> <span class="mi">4</span><span class="p">]</span>

            <span class="n">other_fields</span> <span class="o">=</span> <span class="p">{}</span>
            <span class="n">other_fields</span><span class="p">[</span><span class="s2">&quot;log_ratio&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">log_ratio</span>
            <span class="n">other_fields</span><span class="p">[</span><span class="s2">&quot;log_error&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">log_error</span>
            <span class="n">other_fields</span><span class="p">[</span><span class="s2">&quot;p_value&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">p_value</span>
            <span class="n">other_fields</span><span class="p">[</span><span class="s2">&quot;flag&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">flag</span>
            <span class="n">other_fields</span><span class="p">[</span><span class="s2">&quot;substance&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">substance</span>

            <span class="n">profile</span> <span class="o">=</span> <span class="n">expression_profile</span><span class="p">(</span><span class="n">sample_num</span><span class="p">,</span> <span class="n">gene</span><span class="p">,</span> <span class="n">intensity</span><span class="p">,</span> <span class="n">other_fields</span><span class="p">)</span>
            <span class="n">expression_profiles</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">profile</span><span class="p">)</span>

    <span class="nb">file</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    <span class="k">return</span> <span class="n">expression_profiles</span></div>


<div class="viewcode-block" id="readAllExpressionProfiles"><a class="viewcode-back" href="../process_BC_data.html#process_BC_data.readAllExpressionProfiles">[docs]</a><span class="k">def</span> <span class="nf">readAllExpressionProfiles</span><span class="p">(</span><span class="nb">dir</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Reads all files containing expression data in the given directory</span>

<span class="sd">    :param dir: the directory from where to read all text files containing expression data.</span>
<span class="sd">    :type dir: str</span>

<span class="sd">    :return: a list of expression_profiles formed from all files read in the directory</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">files</span> <span class="o">=</span> <span class="p">[</span><span class="n">filename</span> <span class="k">for</span> <span class="n">filename</span> <span class="ow">in</span> <span class="n">glob</span><span class="o">.</span><span class="n">glob</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">dir</span><span class="p">,</span> <span class="s2">&quot;*.txt&quot;</span><span class="p">))]</span>
    <span class="n">expression_profiles</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="nb">file</span> <span class="ow">in</span> <span class="n">files</span><span class="p">:</span>
        <span class="n">expression_profiles</span> <span class="o">+=</span> <span class="n">readExpressionProfile</span><span class="p">(</span><span class="nb">file</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">expression_profiles</span></div>

<div class="viewcode-block" id="getClinicalData"><a class="viewcode-back" href="../process_BC_data.html#process_BC_data.getClinicalData">[docs]</a><span class="k">def</span> <span class="nf">getClinicalData</span><span class="p">(</span><span class="n">file_name</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Reads clinical information about each patient who gave a gene sample</span>

<span class="sd">    :param file_name: the directory and file containing the patient&#39;s clinical information</span>
<span class="sd">    :type file_name: str</span>

<span class="sd">    :return: a list of clinical_data</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="nb">file</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">file_name</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">,</span> <span class="n">encoding</span><span class="o">=</span><span class="s2">&quot;utf-7&quot;</span><span class="p">)</span>

    <span class="n">clinical_datas</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">rows</span> <span class="o">=</span> <span class="nb">file</span><span class="o">.</span><span class="n">read</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>

    <span class="c1">#ignore first line of labels</span>
    <span class="c1">#has extra blank line so don&#39;t take last row</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">rows</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span><span class="p">):</span>
        <span class="n">row</span> <span class="o">=</span> <span class="n">rows</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>

        <span class="n">sample_num</span>      <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>

        <span class="n">first_series</span>    <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
        <span class="n">posnodes</span>        <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>
        <span class="n">event_meta</span>      <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span>
        <span class="n">event_death</span>     <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="mi">4</span><span class="p">]</span>
        <span class="n">time_survival</span>   <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="mi">5</span><span class="p">]</span>
        <span class="n">time_recur</span>      <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="mi">6</span><span class="p">]</span>
        <span class="n">time_meta</span>       <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="mi">7</span><span class="p">]</span>
        <span class="n">esr1</span>            <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="mi">8</span><span class="p">]</span>
        <span class="n">nih</span>             <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="mi">9</span><span class="p">]</span>
        <span class="n">st_gallen</span>       <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="mi">10</span><span class="p">]</span>
        <span class="n">conserv</span>         <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="mi">11</span><span class="p">]</span>
        <span class="n">c1_from_data</span>    <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="mi">12</span><span class="p">]</span>
        <span class="n">c1_cross_valid</span>  <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="mi">13</span><span class="p">]</span>
        <span class="n">c1_used</span>         <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="mi">14</span><span class="p">]</span>

        <span class="n">other_fields</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">other_fields</span><span class="p">[</span><span class="s2">&quot;first_series&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">first_series</span>
        <span class="n">other_fields</span><span class="p">[</span><span class="s2">&quot;posnodes&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">posnodes</span>
        <span class="n">other_fields</span><span class="p">[</span><span class="s2">&quot;event_meta&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">event_meta</span>
        <span class="n">other_fields</span><span class="p">[</span><span class="s2">&quot;event_meta&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">event_meta</span>
        <span class="n">other_fields</span><span class="p">[</span><span class="s2">&quot;event_death&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">event_death</span>
        <span class="n">other_fields</span><span class="p">[</span><span class="s2">&quot;time_survival&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">time_survival</span>
        <span class="n">other_fields</span><span class="p">[</span><span class="s2">&quot;time_recur&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">time_recur</span>
        <span class="n">other_fields</span><span class="p">[</span><span class="s2">&quot;time_meta&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">time_meta</span>
        <span class="n">other_fields</span><span class="p">[</span><span class="s2">&quot;esr1&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">esr1</span>
        <span class="n">other_fields</span><span class="p">[</span><span class="s2">&quot;nih&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">nih</span>
        <span class="n">other_fields</span><span class="p">[</span><span class="s2">&quot;st_gallen&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">st_gallen</span>
        <span class="n">other_fields</span><span class="p">[</span><span class="s2">&quot;conserv&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">conserv</span>
        <span class="n">other_fields</span><span class="p">[</span><span class="s2">&quot;c1_from_data&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">c1_from_data</span>
        <span class="n">other_fields</span><span class="p">[</span><span class="s2">&quot;c1_cross_valid&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">c1_cross_valid</span>
        <span class="n">other_fields</span><span class="p">[</span><span class="s2">&quot;c1_used&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">c1_used</span>

        <span class="n">data</span> <span class="o">=</span> <span class="n">clinical_data</span><span class="p">(</span><span class="n">sample_num</span><span class="p">,</span> <span class="n">other_fields</span><span class="p">)</span>
        <span class="n">clinical_datas</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">clinical_datas</span></div>
</pre></div>

          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper"><div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../index.html">Documentation overview</a><ul>
  <li><a href="index.html">Module code</a><ul>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2016, Andrew Luo & David Gibbs.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.3.5</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.7</a>
      
    </div>

    

    
  </body>
</html>